src/toil/server/wes/tasks.py

```diff
@@ -0,0 +1,279 @@
+        to be executed. Return that list of shell commands that should be
+        executed in order to complete this workflow run.
+        """
+        # write main workflow file and its secondary input file
+        workflow_url = self.write_workflow(src_url=self.request["workflow_url"])
+        input_json = os.path.join(self.exec_dir, "wes_inputs.json")
+
+        # write input file as JSON
+        with open(input_json, "w") as f:
+            json.dump(self.request["workflow_params"], f)
+
+        # create output directory
+        if not os.path.exists(self.out_dir):
+            os.makedirs(self.out_dir)
+
+        # sort commands
+        options = self.sort_options()
+
+        # construct and return the command to run
+        if self.wf_type == "cwl":
+            command_args = (
+                ["toil-cwl-runner"] + options + [workflow_url, input_json]
+            )
+        elif self.wf_type == "wdl":
+            command_args = (
+                ["toil-wdl-runner"] + options + [workflow_url, input_json]
+            )
+        elif self.wf_type == "py":
+            command_args = ["python", workflow_url] + options
+        else:
+            # This shouldn't happen if this was executed from the Toil WES server.
+            raise RuntimeError(f"Unsupported workflow type: {self.wf_type}")
+
+        return command_args
+
+    def call_cmd(self, cmd: Union[List[str], str], cwd: str) -> int:
+        """
+        Calls a command with Popen. Writes stdout, stderr, and the command to
+        separate files. This is a blocking call.
+
+        :returns: The exit code of the command.
+        """
+        stdout_f = os.path.join(self.work_dir, "stdout")
+        stderr_f = os.path.join(self.work_dir, "stderr")
+
+        with open(stdout_f, "w") as stdout, open(stderr_f, "w") as stderr:
+            logger.info(f"Calling: '{' '.join(cmd)}'")
+            process = subprocess.Popen(cmd, stdout=stdout, stderr=stderr, close_fds=True, cwd=cwd)
+
+        try:
+            return process.wait()
+        except (KeyboardInterrupt, SystemExit):
+            process.terminate()
+
+            if process.wait(timeout=5) is None:
+                process.kill()
+
+            logger.info("Child process terminated by interruption.")
+            return 130
+
+    def run(self) -> None:
+        """
+        Construct a command to run a the requested workflow with the options,
+        run it, and deposit the outputs in the output directory.
+        """
+
+        # the task has been picked up by the runner and is currently preparing to run
+        self.set_state("INITIALIZING")
+        commands = self.initialize_run()
+
+        # store the job store location
+        with open(os.path.join(self.work_dir, "job_store"), "w") as f:
+            f.write(self.job_store)
+
+        if self.get_state() in ("CANCELING", "CANCELED"):
+            logger.info("Workflow canceled.")
+            return
+
+        self.set_state("RUNNING")
```

Comment:
This doesn't actually prevent a transition from `CANCELING` to `RUNNING`; we could be put in the `CANCELING` state between the check and the set.

Reply:
Hmm.. Do you think it would help if we use a sqlite3 database to store and update the states? We could do the check and the set in a transaction. It'll probably fix the atomicity issue above as well. Unless there's a better way to do this in Python.

Reply:
A database with transactions would work, but it would be pretty heavy-weight. You can also lock files in a couple of ways, to synchronize access on them so that processes don't have races on their contents.

Maybe the database is the right way to go, since this is going to be a server and will need to handle a bunch of incoming requests and work on shared server state. But then it would be natural to put more than *just* the state of each workflow in the database. Maybe we'd want to use some higher-level database stuff intended for Python web apps?